contracts/L2TokenReceiverV2.sol

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {OwnableUpgradeable} from "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+
+import {ISwapRouter} from "@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+import {TransferHelper} from "@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol";
+
+import {IL2TokenReceiverV2, IERC165, IERC721Receiver} from "./interfaces/IL2TokenReceiverV2.sol";
+import {INonfungiblePositionManager} from "./interfaces/uniswap-v3/INonfungiblePositionManager.sol";
+
+contract L2TokenReceiverV2 is IL2TokenReceiverV2, OwnableUpgradeable, UUPSUpgradeable {
+    address public router;
+    address public nonfungiblePositionManager;
+
+    SwapParams public params;
+
+    // v2
+
+    SwapParams public initialTokenParams;
+
+    constructor() {
+        _disableInitializers();
+    }
+
+    function L2TokenReceiver__init(
+        address router_,
+        address nonfungiblePositionManager_,
+        SwapParams memory params_
+    ) external initializer {
+        __Ownable_init();
+        __UUPSUpgradeable_init();
+
+        router = router_;
+        nonfungiblePositionManager = nonfungiblePositionManager_;
+
+        _editParams(params_, false);
+    }
+
+    function supportsInterface(bytes4 interfaceId_) external pure returns (bool) {
+        return
+            interfaceId_ == type(IL2TokenReceiverV2).interfaceId ||
+            interfaceId_ == type(IERC721Receiver).interfaceId ||
+            interfaceId_ == type(IERC165).interfaceId;
+    }
+
+    function editParams(SwapParams memory newParams_, bool isInitial_) external onlyOwner {
+        SwapParams memory params_ = isInitial_ ? initialTokenParams : params;
+
+        if (params_.tokenIn != address(0) && params_.tokenIn != newParams_.tokenIn) {
+            TransferHelper.safeApprove(params_.tokenIn, router, 0);
+            TransferHelper.safeApprove(params_.tokenIn, nonfungiblePositionManager, 0);
+        }
+
+        if (params_.tokenIn != address(0) && params_.tokenOut != newParams_.tokenOut) {
```

Comment:
This check appeared because we will initialize the second structure with this function. And it is necessary that we do not take away allowance from the zero address.